### ISSUE TITLE
Strange slowness when training shape-predictor model

### ISSUE THREAD

**Issue by Climax777** [davisking/dlib#1101]
## Expected Behavior
We used [this code](https://github.com/davisking/dlib/blob/master/tools/archive/train_face_5point_model.cpp) that you used for the 5-point facial landmark shape prediction. We've used the 5-point training data and annotated three additional points: left-mouth, right-mouth and chin. We modified the code that flips the dataset to look like:
```
template <
    typename image_array_type,
    typename T
    >
void add_image_left_right_flips_8points (
    image_array_type& images,
    std::vector<std::vector<T> >& objects
)
{
    // make sure requires clause is not broken
    DLIB_ASSERT( images.size() == objects.size(),
        "\t void add_image_left_right_flips()"
        << "\n\t Invalid inputs were given to this function."
        << "\n\t images.size():  " << images.size() 
        << "\n\t objects.size(): " << objects.size() 
        );

    typename image_array_type::value_type temp;
    std::vector<T> rects;

    const unsigned long num = images.size();
    for (unsigned long j = 0; j < num; ++j)
    {
        const point_transform_affine tran = flip_image_left_right(images[j], temp);

        rects.clear();
        for (unsigned long i = 0; i < objects[j].size(); ++i)
        {
            rects.push_back(impl::tform_object(tran, objects[j][i]));

            DLIB_CASSERT(rects.back().num_parts() == 8); //<--This line was changed
            swap(rects.back().part(0), rects.back().part(2));
            swap(rects.back().part(1), rects.back().part(3));
           swap(rects.back().part(5), rects.back().part(6)); //<---We added this line for the mouth
        }

        images.push_back(temp);
        objects.push_back(rects);
    }
}
```

We expected to get ballpark the same training time as for the 5-point (around 30 hours) with the same setup.

## Current Behavior
What we found was that the training was virtually brought a halt (1400 hours estimate). The function that was to blame is found in `randomly_generate_split_feature` in `shape_predictor_trainer.h`. One iteration takes anything between one second to five seconds, which is in a tight loop.

When we comment out the flipping and rotation, the training time is ballpark correct (5 hours). When either flipping or rotation is added it becomes impossibly slow.

We then used `imglab`'s own rotation and flipping (the new one) and used that, which then brought us back to 34 hours.

Also retraining the 5-point with flipping/rotation trains within 30 hours.

## Steps to Reproduce
* Use [this code](https://github.com/davisking/dlib/blob/master/tools/archive/train_face_5point_model.cpp) with the two modifications (cassert for 8 points and flipping the labels of the mouth) also available with the manifests below
* Use the 5-point training data [here](http://dlib.net/files/data/dlib_faces_5points.tar) with [these](https://drive.google.com/drive/folders/18u4dsULeqHRN7Wr9jyNqu1NhPkxkU15S?usp=sharing) xml manifests.
* Observe slowness

* **Version**: 19.9
* **Where did you get dlib**: built from source from github releases
* **Platform**: `Ubuntu 17.10 x64 on Intel i7`
* **Compiler**: `gcc (Ubuntu 7.2.0-8ubuntu3) 7.2.0`

## Remarks
I'm not sure if it is essentially a problem/bug or if there may be something wrong with the flipping/rotation. It might be due to the randomness?

We would also like to contribute the following if you are interested
- The 8-point annotations
- The trained model (imglab flipped/rotated)

**Comment by davisking** [comment]
It's happening because you didn't include annotations for every part.  The data loading routines then set those parts to OBJECT_PART_NOT_PRESENT, which just makes parts with a certain really huge point location.  That would be fine and the training code would do the right thing, except the flipping code in that file that you have doesn't do the right thing.  When it flips points with x,y values of OBJECT_PART_NOT_PRESENT it messes them up and turns them into some really huge different point which then goes on to derail the entire rest of the code.  

You should label all the points.  

I'll also put some error or warning message somewhere to deal with this kind of thing in the future.

**Comment by davisking** [comment]
Oh yeah, and more data sharing is always good :)

**Comment by davisking** [comment]
Actually, just get the latest dlib. I just pushed some changes that should just make this go away and do the right thing.

Although you should still probably review your training data and add those missing parts if that isn't what you meant to do.  It's also generally a good idea to make a pass over the data with the model and look for things in disagreement with the model since those are usually annotation errors.  Refining your training dataset in this kind of iterative way almost always results in superior models in the end.

**Comment by Climax777** [comment]
Yup! You were right (not that I expected different). There was one of the boxes that we missed when duplicating the annotations. 

Should we retrain now? Maybe with different parameters that you could recommend? 

The results are pretty good as is though...

**Comment by davisking** [comment]
I would try a variety of parameters.  I would also try turning on the box relative padding instead of landmark relative padding.  Just try different things and see what happens.  Also note that those parameters in that file are somewhat extreme. You could turn a lot of them down and still get good performance.  I turned them up to make the final model because why not?  But for doing parameter sweeps it's sensible to turn things down to make it run faster, assuming you don't want to just wait.

You can also use this thing to help with the parameter selection: http://blog.dlib.net/2017/12/a-global-optimization-algorithm-worth.html

**Comment by Climax777** [comment]
Thanks. I guess this is resolved now. I will stay in touch to upload the final annotations and model
